pyramid_zipkin/tween.py

```diff
@@ -8,6 +9,72 @@
 from pyramid_zipkin.request_helper import get_binary_annotations
 
 
+ZipkinSettings = namedtuple('ZipkinSettings', [
+    'zipkin_attrs',
+    'transport_handler',
+    'service_name',
+    'span_name',
+    'add_logging_annotation',
+    'report_root_timestamp',
+])
+
+
+def _get_settings_from_request(request):
+    """Extracts Zipkin attributes and configuration from request attributes.
+    See the `zipkin_span` context in py-zipkin for more detaied information on
+    all the settings.
+
+    Here are the supported Pyramid registry settings:
+
+    zipkin.create_zipkin_attr: allows the service to override the creation of
+        Zipkin attributes. For example, if you want to deterministically
+        calculate trace ID from some service-specific attributes.
+    zipkin.transport_handler: how py-zipkin will log the spans it generates.
+    zipkin.stream_name: an additional parameter to be used as the first arg
+        to the transport_handler function. A good example is a Kafka topic.
+    zipkin.add_logging_annotation: if true, the outermost span in this service
+        will have an annotation set when py-zipkin begins its logging.
+    zipkin.report_root_timestamp: if true, the outermost span in this service
```

Comment:
pyramid-zipkin [passes in attributes](https://github.com/Yelp/pyramid_zipkin/blob/master/pyramid_zipkin/tween.py#L52) to py-zipkin's `zipkin_span` regardless of whether it actually extracts headers from the incoming request. This is set up this way to allow the service to do other useful things with the Zipkin attributes, like deterministically calculating X-B3-TraceId based on another unique ID, or deterministically sampling requests based on a unique ID. We do both of those things.

As such, the logic I added in https://github.com/Yelp/py_zipkin/pull/30 doesn't actually work for pyramid-zipkin, since Zipkin attributes are always passed in :).

This is kind of an unfortunate setup, I agree. It's grown out of some hacks we've introduced to allow logging spans in a number of out-of-band situations. We've done this to avoid having to write proper Zipkin modules for HAProxy, Nginx, Apache, and uWSGI.